src/privacy_accounting/optimizer/dp_optimizer.py

```diff
@@ -0,0 +1,100 @@
+from math import ceil
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+from src.privacy_accounting.analysis import QueryWithLedger
+from src.privacy_accounting.dp_query import GaussianDPQuery
+from . import WrapperOptimiser
+
+
+class DPOptimiser(WrapperOptimiser):
+
+    def __init__(self,
+                 optimiser,
+                 model,
+                 vector_loss,
+                 dp_sum_query,
+                 num_microbatches=None):
+
+        self.optimiser = optimiser
+        self.model = model
+        self.vector_loss = vector_loss
+        self.dp_sum_query = dp_sum_query
+        self.num_microbatches = num_microbatches
+
+        self._global_parameters = self.dp_sum_query.initial_global_state()
+
+    def fit_batch(self, x: torch.Tensor, y: torch.Tensor):
+
+        loss = self.vector_loss(self.model(x), y)
+
+        param_groups = self.optimiser.param_groups
+
+        sample_state = self.dp_sum_query.initial_sample_state(
```

Comment:
Basically the optimiser stores a reference to any parameter you want it to optimise. This grabs a reference to all those so that we can work on thier graidents

Reply:
what does the sample state actually store?

Reply:
The current state of the sample being drawn from the dataset. So in microbatch SGD, the accumulated clipped gradients so far

Reply:
add a comment pls